server/stream.go

```diff
@@ -4121,15 +4126,48 @@ func (mset *stream) storeMsgId(dde *ddentry) {
 }
 
 // storeMsgIdLocked will store the message id for duplicate detection.
-// Lock should he held.
+// Lock should be held.
 func (mset *stream) storeMsgIdLocked(dde *ddentry) {
 	if mset.ddmap == nil {
-		mset.ddmap = make(map[string]*ddentry)
+		mset.ddmap = stree.NewSubjectTree[*ddentry]()
 	}
-	mset.ddmap[dde.id] = dde
-	mset.ddarr = append(mset.ddarr, dde)
-	if mset.ddtmr == nil {
-		mset.ddtmr = time.AfterFunc(mset.cfg.Duplicates, mset.purgeMsgIds)
+
+	// Clean up the previous entry we just replaced. Stitch the gap.
+	if o, updated := mset.ddmap.Insert(stringToBytes(dde.id), dde); updated && o != nil {
```

Comment:
If updated is true that means we had a collision yes? Should this not happen?

Reply:
Meaning we should only put new items in here and the msgID should have been checked prior?

Reply:
I've added a comment to the code to this effect. Ordinarily yes, `checkMsgId()` has been called by now, but when we have to call `rebuildDedupe()`, we will replay all of the stream messages in the deduplication window and they may have duplicate message IDs. Therefore we only want to keep the last `ddentry` for each ID.

Reply:
I think the only way it should normally happen is if the duplicate window was changed before calling `rebuildDedupe()`.